src/network/cni/namespace.rs

```diff
@@ -0,0 +1,61 @@
+//! Network namespace helpers and structures.
+
+use anyhow::{Context, Error, Result};
+use nix::sched::{setns, CloneFlags};
+use std::{future::Future, os::unix::io::AsRawFd, path::Path};
+use tokio::fs::File;
+
+#[derive(Debug)]
+/// A basic network namespace abstraction.
+pub struct Namespace {
+    file: File,
+}
+
+impl Namespace {
+    /// Create a new namespace.
+    pub async fn new<P>(path: P) -> Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        Ok(Self {
+            file: File::open(&path).await.context("open namespace file")?,
+        })
+    }
+
+    /// Run a closure inside the namespace
+    pub async fn run<F>(&self, fun: F) -> Result<()>
+    where
+        F: Future<Output = Result<()>> + Send + 'static,
+    {
+        let fd = self.file.as_raw_fd();
+        tokio::spawn(async move {
+            setns(fd, CloneFlags::CLONE_NEWNET).context("switch to network namespace")?;
+
+            fun.await.context("run namespace future")?;
+
+            let ok: std::result::Result<(), Error> = Ok(());
+            ok
+        })
+        .await
+        .context("spawn namespace thread")?
+        .context("run in namespace thread")
+    }
+}
```

Comment:
@mrunalp I'm really wondering if this is the right implementation for running a future inside of a network namespace (path).

Reply:
Tested and it works. 🙂